fs-index/src/index.rs

```diff
@@ -1,1120 +1,436 @@
+        &self,
+        id: &Id,
+    ) -> Option<&Vec<IndexedResource<Id>>> {
+        self.id_to_resources.get(id)
     }
 
-    #[test]
-    fn update_one_should_error_on_new_file() {
-        run_test_and_clean_up(|path| {
-            create_file_at(path.clone(), Some(FILE_SIZE_1), None);
-            let mut index = ResourceIndex::build(path.clone());
+    /// Get a resource by its path
+    ///
+    /// Returns None if the resource does not exist
+    ///
+    /// **Note**: The path should be relative to the root path
+    pub fn get_resource_by_path<P: AsRef<Path>>(
+        &self,
+        path: P,
+    ) -> Option<&IndexedResource<Id>> {
+        self.path_to_resource.get(path.as_ref())
+    }
+
+    /// Build a new index from the given root path
+    pub fn build<P: AsRef<Path>>(root_path: P) -> Result<Self> {
+        log::debug!("Building index at root path: {:?}", root_path.as_ref());
+
+        // Canonicalize the root path
+        let root = fs::canonicalize(&root_path)?;
+        let mut id_to_resources = HashMap::new();
+        let mut path_to_resource = HashMap::new();
+
+        // Discover paths in the root directory
+        let paths = discover_paths(&root)?;
+        let entries: HashMap<PathBuf, IndexedResource<Id>> =
+            scan_entries(paths);
+
+        // Strip the root path from the entries
+        let entries: HashMap<PathBuf, IndexedResource<Id>> = entries
+            .into_iter()
+            .map(|(path, resource)| {
+                let relative_path =
+                    path.strip_prefix(&root).unwrap().to_path_buf();
+                let resource = IndexedResource::new(
+                    resource.id().clone(),
+                    relative_path.clone(),
+                    resource.last_modified(),
+                );
+                (relative_path, resource)
+            })
+            .collect();
 
-            let (_, new_path) =
-                create_file_at(path.clone(), Some(FILE_SIZE_2), None);
+        // Update the path to resource map
+        path_to_resource.extend(entries.clone());
 
-            let update = index.update_one(&new_path, CRC32_2);
+        // Update the ID to resources map
+        for resource in entries.values() {
+            let id = resource.id().clone();
+            id_to_resources
+                .entry(id)
+                .or_insert_with(Vec::new)
+                .push(resource.clone());
+        }
 
-            assert!(update.is_err())
-        })
+        let index = ResourceIndex {
+            root,
+            id_to_resources,
+            path_to_resource,
+        };
+        Ok(index)
     }
 
-    #[test]
-    fn update_one_should_index_delete_file_successfully() {
-        run_test_and_clean_up(|path| {
-            create_file_at(path.clone(), Some(FILE_SIZE_1), Some(FILE_NAME_1));
+    /// Update the index with the latest information from the file system
+    pub fn update_all(&mut self) -> Result<IndexUpdate<Id>> {
```

Comment:
@kirillt compare with the original implementation

Reply:
I looked into the CI error for Windows in this PR and found that the test case also fails for Windows in `arklib`. It seems the logic in `update_all()` needs another review and debugging to understand why it only fails on Windows.

You can see the CI results for `arklib` from my fork here:
- [Comparison with main](https://github.com/ARK-Builders/arklib/compare/main...tareknaser:arklib:main)
- [CI run details](https://github.com/tareknaser/arklib/actions/runs/10011315002)